Move completed transaction survey from frontend to feedback

Rosa-Fox opened the pull request:
this will need to be deployed with a branch from [Frontend](https://github.com/alphagov/frontend/pull/3534) (still need to remove the translations and update the README on there!) and [Publisher](https://github.com/alphagov/publisher/compare/move-completed-transaction-pages-frontend-to-feedback).

[Trello](https://trello.com/c/VHZ1R1zY/262-no-form-validation-on-service-feedback-forms-l)

# Aim of this work

Completed transaction ‘done’ pages have a url of `/done/transaction-name`. They are **published** from the Publisher application. Here are some examples:

<img width="718" alt="completed-transaction-done-publisher" src="https://user-images.githubusercontent.com/5963488/216365893-e80a51da-cc5e-4ef4-bfe4-c6d456ea06a5.png"> 

The aim of this change is to move the **rendering** of completed_transaction ‘done’ pages from the Frontend application to the Feedback application. Previously the views were rendered in the Frontend application, but the data was [posted to a route](https://github.com/alphagov/frontend/blob/main/app/views/completed_transaction/_standard_satisfaction_survey.html.erb#L6) in the Feedback app which was defined [here](https://github.com/alphagov/feedback/blob/main/config/routes.rb#L20) before this PR.

The problem with this was that there was no form validation, as the errors couldn’t be re-rendered between applications. Given that the views rendered based on data from the content store and the data is posted to the Feedback application, it made sense to move the completed_transaction code out of Frontend and into Feedback.

# Types of Completed Transaction feedback forms

### Service Feedback
Most of the /done/completed-transaction pages render a Service Feedback form. An example is [www.gov.uk/done/vehicle-tax](http://www.gov.uk/done/vehicle-tax). 

### Assisted Digital
There are also three assisted digital feedback forms:

https://www.gov.uk/done/register-flood-risk-exemption
https://www.gov.uk/done/waste-carrier-or-broker-registration
https://www.gov.uk/done/register-waste-exemption

In this PR, they are sent to the appropriate controller using a routing constraint. 

# Transaction finished

https://www.gov.uk/done/transaction-finished
This doesn’t display a form… just content to say the transaction is finished.

# Where is the data sent?

The Service Feedback form fields also exist within the Assisted Digital Feedback form. They are foundationally the same, but Assisted Digital Feedback has some extra fields.

The Service Feedback form data from both forms is sent to the Support API. It can be viewed using the Support app within the Feedback Explorer. Therefore this app does not have its own database and does not store data, it just passes it through. 

In addition to sending some data to the Support API, the data from the other fields plus some data from hidden fields appended using JS (referrer and javascript_enabled) are written to a Google spreadsheet.

# Rendering data from the Content Item

This functionality needed to be added in. It uses the base path to find the content item in the Content Store and then presents the necessary information on the views, which is a common GOV.UK pattern.

# Validations

Conveniently the model validations had already been written into the ServiceFeedback and AssistedDigitialFeedback models, they just weren’t actually rendered before. I changed the order of a few of them to make the messaging more user friendly. 

The error message data is stored within instance variables generated based on the field names. I took this approach as it moves logic from the views to the controllers making the views cleaner. 

# Translations
I added some functionality to check the locale in the content item and then render the translation accordingly. Translations are English and Welsh, en by default. I am currently waiting on the Welsh translations for the validation errors and for some of the promotional content. This will be updated in due course.

# Testing locally

I couldn’t actually run posting to the Support API locally so I tested that on integration.

For testing the form validations try some service feedback pages such as:

http://feedback.dev.gov.uk/done/vehicle-tax

and assisted digital feedback pages such as:

http://feedback.dev.gov.uk/done/register-flood-risk-exemption

A welsh page:

http://feedback.dev.gov.uk/done/ceisio-am-lun-id-tystysgrif-awdurdod-pleidleiswyr

Checking the styling:

Rendering the step by step nav and header:
http://feedback.dev.gov.uk/done/find-pension-contact-details

Rendering the footer:
http://feedback.dev.gov.uk/done/vehicle-tax


Rendering the side nav: 
https://www.gov.uk/done/transaction-finished


For submitting the Assisted Digital Feedback form you will need to get Google API credentials from the integration server and add them to a dotenv file. I can help with this and also finding the spreadsheet. 

Testing on integration also involves deploying a [branch of Frontend](https://github.com/alphagov/frontend/compare/remove-completed-transaction-done-forms) (which removes the routing) and a branch of  [Publisher](https://github.com/alphagov/publisher/compare/move-completed-transaction-pages-frontend-to-feedback) (to switch the rendering app) and republishing a document. I will put together a deploy plan, but it will just involve deploying those PRs and republishing the completed transactions. There is a pre-existing rake task in Publisher to republish per format which I plan to use. 

# Associated PRs:
[Frontend](https://github.com/alphagov/frontend/pull/3534) 
[Publisher](https://github.com/alphagov/publisher/compare/move-completed-transaction-pages-frontend-to-feedback)

# Screen grabs from testing on Integration - please not that the styling has been updated a little bit since this.
(Click on the them to see larger image)

## Service Feedback Form

| Rendered      | Validation errors | Content Item | Support App | 
| ----------- | ----------- | ----------- | -----------|
| <img width="1218" alt="service-feedback-rendered" src="https://user-images.githubusercontent.com/5963488/216371634-4e5481f7-dd28-40f8-874b-a84cca3676b8.png">      | <img width="532" alt="service-feedback-validation-errors-local" src="https://user-images.githubusercontent.com/5963488/216371521-000f2d03-2bd5-4d6d-8fb0-44bf22bdda7b.png">       |  <img width="759" alt="service-feedback-content-item" src="https://user-images.githubusercontent.com/5963488/216371771-5be20695-06fa-4978-986a-6819b7822723.png">       | <img width="1323" alt="service-feedback-support-app" src="https://user-images.githubusercontent.com/5963488/216371843-3f304f6a-1b1e-4ae4-bd68-37cff1a4e087.png">       |

| Rendered      | Validation errors | Content Item | Support App | Spreadsheet | 
| ----------- | ----------- | ----------- | ----------- |  ----------- |
| <img width="353" alt="assisted-digital-feedback-rendered" src="https://user-images.githubusercontent.com/5963488/216372851-e91e0e53-0a2c-47ee-bf9b-24d59c460b2e.png">      | <img width="948" alt="assisted-digital-feedback-validations" src="https://user-images.githubusercontent.com/5963488/216372925-f13033ea-d97a-4fc1-b649-be3c08959ffa.png">      | <img width="792" alt="assisted-digital-feedback-content-item" src="https://user-images.githubusercontent.com/5963488/216372994-381642f1-6d46-4f76-9862-43f13110de8d.png">      |  <img width="1304" alt="assisted-digital-feedback-support-app" src="https://user-images.githubusercontent.com/5963488/216373056-87449a7d-da0b-4aa6-9610-38ae8d280fcb.png">      |  ![assisted-digital-feedback-spreadsheet-1](https://user-images.githubusercontent.com/5963488/216373689-0eeb799e-c0d0-427b-a704-0d91574c334d.png) |


Welsh page

| Rendered      | Validation errors | Content Item | Support App | 
| ----------- | ----------- | ----------- | -----------|
|   <img width="1140" alt="welsh-page-rendered" src="https://user-images.githubusercontent.com/5963488/216373996-6afa9041-8a3c-44ad-9491-4c5886fed118.png"> |  <img width="950" alt="welsh-page-validation" src="https://user-images.githubusercontent.com/5963488/216374050-66086869-c2ca-4e36-a25e-ba4d5ee12613.png">  |  <img width="754" alt="welsh-page-content-item" src="https://user-images.githubusercontent.com/5963488/216374119-f78dcd43-1b39-4cc6-a2ab-864b96adcab7.png">  | <img width="1350" alt="welsh-page-support-app" src="https://user-images.githubusercontent.com/5963488/216374184-ccdcb133-dbe2-4a9b-8e2b-9c4ef30a6589.png">   |


⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

gclssvglx commented on spec/lib/format_routing_constraint_spec.rb:
Lovely tests, thanks @Rosa-Fox

gclssvglx commented on spec/presenters/content_item_presenter_spec.rb:
Fantastic! These are super helpful 🥇

Rosa-Fox replied:
👍

gclssvglx commented on app/controllers/service_feedback_controller.rb:
```diff
@@ -11,4 +11,62 @@ def new
+  def confirm_submission
+    redirect_to contact_anonymous_feedback_thankyou_path
+  end
+  def set_form_field_values
```
Might need a space here I think.

gclssvglx commented on app/views/shared/_error_messages.html.erb:
Nice!

gclssvglx commented on spec/support/service_feedback.rb:
Great stuff!

gclssvglx commented on app/views/shared/_completed_transaction_feedback.html.erb:
```diff
@@ -0,0 +1,58 @@
+    <p class="govuk-body"><%= t("controllers.contact.govuk.service_feedback.thanks_for_visiting") %></p>
+  <% end %>
+  <% if publication.promotion %>
+    <div class="promotion">
```
I wonder if a [Notice](https://components.publishing.service.gov.uk/component-guide/notice) would work here 🤔

Rosa-Fox replied:
ah I was just copying the previous design. I could ask about it and do as a separate PR?

gclssvglx commented on app/models/assisted_digital_feedback.rb:
```diff
@@ -112,11 +130,27 @@ def assistance_provided_by_other?
     !url.nil? ? URI(url).path : nil
   end
+
+  def assistance_satisfaction_rating_row_data
+    if assistance_provided_by_government_staff?
+      assistance_satisfaction_rating.to_i
+    elsif assistance_provided_by_other?
+      assistance_satisfaction_rating_other.to_i
+    end
+  end
+
+  def assistance_improvement_comments_row_data
+    if assistance_provided_by_government_staff?
+      assistance_improvement_comments
+    elsif assistance_provided_by_other?
+      assistance_improvement_comments_other
+    end
```
Cool 👍

gclssvglx commented on app/views/assisted_digital_feedback/_assistance_received_other.html.erb:
```diff
@@ -0,0 +1,63 @@
+      value: @value_assistance_improvement_comments_other
+    },
+    maxlength: 1250
+  } %>
```
Might be missing an empty last line I think

gclssvglx commented on app/views/assisted_digital_feedback/_assisted_digital_satisfaction_survey.html.erb:
```diff
@@ -0,0 +1,141 @@
+      maxlength: 1250
+    } %>
+
+    <%= render "govuk_publishing_components/components/radio", {
+      name: "service_feedback[assistance_provided_by]",
+      heading: t("controllers.contact.govuk.assisted_digital_feedback.who_assisted"),
+      heading_size: "s",
+      error_message: @error_message_assistance_provided_by,
+      id: @error_id_assistance_provided_by,
+      items: [
+        {
+          value: "friend-relative",
+          text: t("controllers.contact.govuk.assisted_digital_feedback.friend_or_relative"),
+          checked: @value_assistance_provided_by == "friend-relative"
+        },
+        {
+          value: "work-colleague",
+          text: t("controllers.contact.govuk.assisted_digital_feedback.work_colleague"),
+          checked: @value_assistance_provided_by == "work-colleague"
+        },
+        {
+          value: "government-staff",
+          text: t("controllers.contact.govuk.assisted_digital_feedback.government_staff"),
+          checked: @value_assistance_provided_by == "government-staff",
+          conditional: government_staff
+        },
+        {
+          value: "other",
+          text: t("controllers.contact.govuk.assisted_digital_feedback.other"),
+          checked: @value_assistance_provided_by == "other",
+          conditional: other_person
+        },
+      ]
+    } %>
+  <% end %>
```
I wonder if we could turn this into a partial - feel free to say no if that won't help?

Rosa-Fox replied:
ah I kept it in one file as I thought it would be easier to read with the captures.

gclssvglx replied:
No worries

gclssvglx commented on spec/requests/assisted_digital_spec.rb:
🥇

gclssvglx reviewed (commented):
This is amazing work @Rosa-Fox! The scale of these changes is really substantial - great stuff working all this out. Super impressive and I'm totally in awe of your tests 🥇 

I know you're not quite finished with yet @Rosa-Fox, so, but I thought it'd be useful to see some of my comments as some are easy fixes and some might help remove some duplication. No blockers in here though.

Rosa-Fox commented on config/routes.rb:
```diff
@@ -6,6 +6,18 @@
   get "/contact", format: false, to: "contact#index"
 
+  constraints FormatRoutingConstraint.new("completed_transaction") do
+    get "*slug", slug: %r{done/register-flood-risk-exemption}, to: "assisted_digital_feedback#new"
+    post "*slug", slug: %r{done/register-flood-risk-exemption}, to: "assisted_digital_feedback#create"
+    get "*slug", slug: %r{done/waste-carrier-or-broker-registration}, to: "assisted_digital_feedback#new"
+    post "*slug", slug: %r{done/waste-carrier-or-broker-registration}, to: "assisted_digital_feedback#create"
+    get "*slug", slug: %r{done/register-waste-exemption}, to: "assisted_digital_feedback#new"
+    post "*slug", slug: %r{done/register-waste-exemption}, to: "assisted_digital_feedback#create"
```
ah /done clashes with the service_feedback URLs. I couldn't think of a regex that would catch these 3 and they can be hard to read... but appreciate it is a bit weird

gclssvglx replied:
Argh! That's a shame. Let's not worry for now.

Rosa-Fox commented on app/controllers/service_feedback_controller.rb:
```diff
@@ -11,4 +11,62 @@ def new
+
+      set_form_field_values
+      set_error_message_per_component
+      set_error_message_id_per_component
```
I added a method to encapsulate handling the form errors as a new commit.

gclssvglx replied:
Perfect - thanks Rosa

Rosa-Fox commented on app/controllers/assisted_digital_feedback_controller.rb:
```diff
@@ -0,0 +1,75 @@
+      @publication = helpers.publication
+      set_form_field_values
+      set_error_message_per_component
+      set_error_message_id_per_component
```
added as a new commit

gclssvglx replied:
Thanks

Rosa-Fox commented:
@gclssvglx this is ready for review again when you are back. I added 2 new commits, one is because the way that loading the styles had changed when I rebased.

gclssvglx reviewed (approved):
Great work @Rosa-Fox
